commit ee2d61df273e85a4855ddf6a8af5f137109b8b91 (refs/changes/79/2079/2)
Author: Jordan Hendricks <jordan.hendricks@joyent.com>
Date:   2017-06-13T13:25:40-07:00 (2 years, 4 months ago)
    
    joyent/node-manta#302 Create a manual page for `mmpu`
    joyent/node-manta#311 `createUpload` incorrectly handles some target object headers
    Reviewed by:
    Approved by:

diff --git a/CHANGES.md b/CHANGES.md
index ac710a4..201ee1a 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -2,6 +2,10 @@
 
 ## not yet released
 
+## 4.4.1
+- joyent/node-manta#302 Create a manual page for `mmpu`
+- joyent/node-manta#311 `createUpload` incorrectly handles some target object headers
+
 ## 4.4.0
 
 - joyent/node-manta#308 `mmpu commit` does not parse options
diff --git a/bin/mmpu b/bin/mmpu
index 7f27650..b01b92d 100755
--- a/bin/mmpu
+++ b/bin/mmpu
@@ -138,7 +138,7 @@ MMpu.prototype.do_create = function do_create(subcmd, opts, args, cb) {
 
     var objectPath = args[0];
     var client = createClient(opts);
-    var headers = {};
+    var targetObjectHeaders = {};
 
     (opts.header || []).forEach(function (h) {
         if (h.indexOf(':') === -1) {
@@ -146,11 +146,11 @@ MMpu.prototype.do_create = function do_create(subcmd, opts, args, cb) {
             return;
         }
         var tmp = strsplit(h, ':', 2);
-        headers[tmp[0]] = tmp[1].trim();
+        targetObjectHeaders[tmp[0]] = tmp[1].trim();
     });
 
     var createUploadOpts = {
-        headers: headers,
+        targetObjectHeaders: targetObjectHeaders,
         account: opts.account,
         md5: opts.md5,
         copies: opts.copies,
@@ -205,7 +205,6 @@ MMpu.prototype.do_create.options = manta.DEFAULT_CLI_OPTIONS.concat([
 ]);
 
 
-//TODO: add some examples here.
 MMpu.prototype.do_create.help = [
     'Initiate a Manta multipart upload for an object.',
     '',
@@ -466,7 +465,6 @@ MMpu.prototype.do_commit.help = [
     '',
     '   $ mmpu commit be5f05d6-7daa-6869-d19b-c1f7fbcf6d8d \\',
     '        f14a41dc-7d28-6fdb-e07c-d54c0adcdf35',
-    '        e2893e52-9ba3-64fe-eec9-b4663835ad01',
     '',
     '   $ mmpu commit be5f05d6-7daa-6869-d19b-c1f7fbcf6d8d \\',
     '        f14a41dc-7d28-6fdb-e07c-d54c0adcdf35 \\',
diff --git a/docs/man/mchattr.md b/docs/man/mchattr.md
index e44c545..c8388d6 100644
--- a/docs/man/mchattr.md
+++ b/docs/man/mchattr.md
@@ -64,7 +64,7 @@ ENVIRONMENT
 -----------
 
 `MANTA_USER`
-  In place of `-a, --account`
+  In place of `-a, --account`.
 
 `MANTA_SUBUSER`
   In place of `--user`.
diff --git a/docs/man/mchmod.md b/docs/man/mchmod.md
index f08f8e4..26e0b33 100644
--- a/docs/man/mchmod.md
+++ b/docs/man/mchmod.md
@@ -66,7 +66,7 @@ ENVIRONMENT
 -----------
 
 `MANTA_USER`
-  In place of `-a, --account`
+  In place of `-a, --account`.
 
 `MANTA_SUBUSER`
   In place of `--user`.
diff --git a/docs/man/mfind.md b/docs/man/mfind.md
index eca420f..22da707 100644
--- a/docs/man/mfind.md
+++ b/docs/man/mfind.md
@@ -99,7 +99,7 @@ OPTIONS
 ENVIRONMENT
 -----------
 `MANTA_USER`
-  In place of `-a, --account`
+  In place of `-a, --account`.
 
 `MANTA_SUBUSER`
   In place of `--user`.
diff --git a/docs/man/mget.md b/docs/man/mget.md
index d3500d6..7187fe0 100644
--- a/docs/man/mget.md
+++ b/docs/man/mget.md
@@ -85,7 +85,7 @@ OPTIONS
 ENVIRONMENT
 -----------
 `MANTA_USER`
-  In place of `-a, --account`
+  In place of `-a, --account`.
 
 `MANTA_SUBUSER`
   In place of `--user`.
diff --git a/docs/man/minfo.md b/docs/man/minfo.md
index 3c8aaeb..cc0eb6b 100644
--- a/docs/man/minfo.md
+++ b/docs/man/minfo.md
@@ -58,10 +58,10 @@ ENVIRONMENT
 -----------
 
 `MANTA_USER`
-  In place of `-a, --account`
+  In place of `-a, --account`.
 
 `MANTA_SUBUSER`
-  In place of `--user`
+  In place of `--user`.
 
 `MANTA_KEY_ID`
   In place of `-k, --key`.
diff --git a/docs/man/mjob.md b/docs/man/mjob.md
index fd3ca00..e4dd8dc 100644
--- a/docs/man/mjob.md
+++ b/docs/man/mjob.md
@@ -317,10 +317,10 @@ ENVIRONMENT
 -----------
 
 `MANTA_USER`
-  In place of `-a, --account`
+  In place of `-a, --account`.
 
 `MANTA_SUBUSER`
-  In place of `--user`
+  In place of `--user`.
 
 `MANTA_KEY_ID`
   In place of `-k, --key`.
diff --git a/docs/man/mln.md b/docs/man/mln.md
index f4feb35..2ebffd8 100644
--- a/docs/man/mln.md
+++ b/docs/man/mln.md
@@ -64,10 +64,10 @@ ENVIRONMENT
 -----------
 
 `MANTA_USER`
-  In place of `-a, --account`
+  In place of `-a, --account`.
 
 `MANTA_SUBUSER`
-  In place of `--user`
+  In place of `--user`.
 
 `MANTA_KEY_ID`
   In place of `-k, --key`.
diff --git a/docs/man/mlogin.md b/docs/man/mlogin.md
index 11af393..317bec0 100644
--- a/docs/man/mlogin.md
+++ b/docs/man/mlogin.md
@@ -132,10 +132,10 @@ ENVIRONMENT
 -----------
 
 `MANTA_USER`
-  In place of `-a, --account`
+  In place of `-a, --account`.
 
 `MANTA_SUBUSER`
-  In place of `--user`
+  In place of `--user`.
 
 `MANTA_KEY_ID`
   In place of `-k, --key`.
diff --git a/docs/man/mls.md b/docs/man/mls.md
index 663258f..d40ecef 100644
--- a/docs/man/mls.md
+++ b/docs/man/mls.md
@@ -78,7 +78,7 @@ ENVIRONMENT
 -----------
 
 `MANTA_USER`
-  In place of `-a, --account`
+  In place of `-a, --account`.
 
 `MANTA_SUBUSER`
   In place of `--user`.
diff --git a/docs/man/mmkdir.md b/docs/man/mmkdir.md
index c1cf545..6b5fe32 100644
--- a/docs/man/mmkdir.md
+++ b/docs/man/mmkdir.md
@@ -61,7 +61,7 @@ ENVIRONMENT
 -----------
 
 `MANTA_USER`
-  In place of `-a, --account`
+  In place of `-a, --account`.
 
 `MANTA_SUBUSER`
   In place of `--user`.
diff --git a/docs/man/mmpu.md b/docs/man/mmpu.md
new file mode 100644
index 0000000..3ba124f
--- /dev/null
+++ b/docs/man/mmpu.md
@@ -0,0 +1,267 @@
+mmpu 1 "June 2017" Manta "Manta Commands"
+=======================================
+
+NAME
+----
+
+mmpu - manage manta multipart uploads
+
+SYNOPSIS
+--------
+
+`mmpu` [OPTION...] command [command-specific arguments]
+
+DESCRIPTION
+-----------
+
+mmpu allows you to interact with Manta's multipart upload API. The multipart
+upload API allows you to upload an object to Manta in chunks, or "parts",
+instead of through a single PUT request.  After all parts have been uploaded,
+you can "commit" your multipart upload, exposing your uploaded object as a
+single file in Manta.  Using mmpu, you can create multipart uploads, upload
+parts to them, and commit or "abort" (cancel) them. You may also list
+non-garbage collected uploads and see which parts for a given upload.
+
+The primary reference for a multipart upload is its UUID.  Most commands operate
+on multipart uploads by UUID.  Parts of a given upload are referenced by their
+part number.
+
+The final object created by a committed multipart upload is referred to as the
+"target object" of the multipart upload.  Additionally, a multipart upload is
+often referred to as an "MPU" or just an "upload".
+
+COMMON OPTIONS
+--------------
+
+The following options are supported in all commands:
+
+`-a, --account login`
+  Authenticate as account (login name).
+
+`-h, --help`
+  Print a help message and exit.
+
+`-i, --insecure`
+  This option explicitly allows "insecure" SSL connections and transfers.  All
+  SSL connections are attempted to be made secure by using the CA certificate
+  bundle installed by default.
+
+`-k, --key fingerprint`
+  Authenticate using the SSH key described by `fingerprint`.  The key must
+  either be in `~/.ssh` or loaded in the SSH agent via `ssh-add`.
+
+`--role=ROLE,ROLE,...`
+  Specify which roles to assume for the request.
+
+`--user user`
+  Authenticate as user under account.
+
+`-u, --url url`
+  Manta base URL (such as `https://manta.us-east.joyent.com`).
+
+`-v, --verbose`
+  Print debug output to stderr.  Repeat option to increase verbosity.
+
+COMMANDS
+--------
+
+The following commands and options are supported:
+
+### create [OPTIONS...] PATH ###
+
+Creates a multipart upload, that, when committed, will be exposed as a Manta
+object at PATH.  When this command completes successfully, it prints a
+multipart upload UUID to stdout.  You should save this UUID for use in other
+commands.
+
+For example, if you wanted to create a multipart upload for which the target
+object will be stored at ~~/stor/foo.txt, you would run:
+
+    $ mmpu create ~~/stor/foo.txt
+
+Since objects created through the multipart upload API should be
+indistinguishable from objects created through a normal Manta PUT, the `create`
+command supports similar options that you might use for `mput`.  For example,
+to create a multipart upload that has 3 copies of the target object, and stores
+some additional headers on the target object metadata, you would run:
+
+    $ mmpu create ~~/stor/foo.txt -c 3 -H 'content-type: text/plain' \
+                                       -H 'access-control-allow-origin: *'
+
+You may also add additional data integrity checks that will be validated
+for your target object when you commit your upload.  For example, if you want
+to verify that your target object ~~/stor/foo.txt is 15728640 bytes and has
+the content-md5 value "FLFyNOI3UFQhtkkrjXV1Bw==", you would run:
+
+    $ mmpu create ~~/stor/foo.txt -s 15728650 \
+                                  -m 'FLFyNOI3UFQhtkkrjXV1Bw=='
+
+The following options are supported on `create`:
+
+`-c, --copies num_copies`
+  Specify the number of copies of the target object to store.
+
+`-H, --header header`
+  Specify a HTTP header to store in the target object's metadata.  As with all
+  Manta objects, you may store up to 4 KB of headers on the target object.
+
+`-m, --md5 md5`
+  Validate the `content-md5` of the target object against the given md5 string.
+  If the MD5 of the target object differs from the provided MD5 when the object
+  is committed, the commit will fail.
+
+  Note that the behavior of this option is different the similar option for `mput`.
+
+`-s, --size size`
+  Validate the size of the target object against `size`.  If the size of the
+  target object differs from this size when the object is committed, the commit
+  will fail.
+
+### upload [-f file] [OPTIONS...] MPU PART_NUM ###
+
+Uploads the contents of stdin, or from the file specified with `-f`, to the
+specified part for a given upload.
+
+Note that parts are zero-indexed.  For example, if you wanted to upload the
+first part of upload be5f05d6-7daa-6869-d19b-c1f7fbcf6d8d, and the part's
+contents are stored locally at ./part0.txt, you could run the following
+equivalent commands:
+
+    $ cat part0.txt | mmpu upload be5f05d6-7daa-6869-d19b-c1f7fbcf6d8d 0
+
+Or:
+
+    $ mmpu upload -f part0.txt be5f05d6-7daa-6869-d19b-c1f7fbcf6d8d 0
+
+When this command completes successfully, it prints the etag of the part
+that was uploaded to stdout.  You should save this etag to use when committing
+the upload.
+
+The following options are supported on `upload`:
+
+`-f, --file file`
+  Upload the contents of `file` as the part, instead of the contents of stdin.
+
+`--progress`
+  Force the progress bar to draw, even when stderr is redirected.
+
+`-q, --quiet`
+  Do not display a progress meter.
+
+### commit MPU [ETAG...] ###
+
+Commits a given upload with the parts specified.  The parts are represented by
+their position in the argument list, which indicates their part number (with
+part 0 being the first argument after the upload ID, the second part the next
+argument after that, and so on), and their contents are represented by their
+etag.  If the etag provided for a part does not match the etag of the part when
+the commit process fetches its metadata, an error will be returned.
+
+For example, if you want to commit an upload with id
+be5f05d6-7daa-6869-d19b-c1f7fbcf6d8d and the following parts:
+    part 0: etag f14a41dc-7d28-6fdb-e07c-d54c0adcdf35
+    part 1: etag e2893e52-9ba3-64fe-eec9-b4663835ad01
+    part 2: etag 73b44fa7-fbd3-efea-b7d0-cd8098e1d928
+
+You would run:
+    $ mmpu commit be5f05d6-7daa-6869-d19b-c1f7fbcf6d8d \
+                  f14a41dc-7d28-6fdb-e07c-d54c0adcdf35 \
+                  e2893e52-9ba3-64fe-eec9-b4663835ad01 \
+                  73b44fa7-fbd3-efea-b7d0-cd8098e1d928
+
+The multipart upload API does not require that you commit all parts uploaded,
+as long as zero or more consecutive parts are committed, and all parts meet
+part size restraints.  So the following commands committing the same upload as
+above are also valid, but would create a different target object than the
+first example.
+
+This command would commit a zero-byte target object:
+    $ mmpu commit be5f05d6-7daa-6869-d19b-c1f7fbcf6d8d
+
+This would create an object consisting of only the first two parts:
+    $ mmpu commit be5f05d6-7daa-6869-d19b-c1f7fbcf6d8d \
+                  f14a41dc-7d28-6fdb-e07c-d54c0adcdf35 \
+                  e2893e52-9ba3-64fe-eec9-b4663835ad01
+
+Once the commit process has begun for a given MPU, it cannot be aborted, or
+committed with a different set of parts than it was initially committed with,
+but you may retry the commit if needed.
+
+To see the current status of an upload, use the `get` command.
+
+### abort MPU ###
+
+Aborts a multipart upload.
+
+Once an upload has begun aborting, it may not be committed, but you may retry
+the abort operation if needed.  To see the current status of an upload, use the
+`get` command.
+
+### get MPU ###
+
+Fetches a JSON blob information about an upload, including its status: created,
+finalizing, or done (committed or aborted).  A finalizing upload is one
+that is either in the process of being committed or aborted, or a commit or
+abort request began but failed.
+
+### list ###
+
+Lists all multipart uploads for a user (note, this can also be done with a
+normal `mfind` call).  Note that these uploads may be in any state; their
+presence merely indicates they have not been garbage collected yet.
+
+If you wish to list all parts that have been uploaded as well, you can run:
+
+    $ mmpu list -p
+
+To see the parts of only one upload, you should use the `parts` command.
+
+The following options are supported on `list`:
+
+`-p, --includeParts`
+  List parts in additional to uploads.
+
+### parts MPU ###
+
+Lists all parts that have been uploaded to a given multipart upload.
+
+
+ENVIRONMENT
+-----------
+
+`MANTA_USER`
+  In place of `-a, --account`.
+
+`MANTA_SUBUSER`
+  In place of `--user`.
+
+`MANTA_KEY_ID`
+  In place of `-k, --key`.
+
+`MANTA_ROLE`
+  In place of `--role`.
+
+`MANTA_URL`
+  In place of `-u, --url`.
+
+`MANTA_TLS_INSECURE`
+  In place of `-i, --insecure`.
+
+The shortcut `~~` is equivalent to `/:login`
+where `:login` is the account login name.
+
+DIAGNOSTICS
+-----------
+
+When using the `-v` option, diagnostics will be sent to stderr in bunyan
+output format.  As an example of tracing all information about a request,
+try:
+
+    $ mmpu create -vv ~~/stor/foo 2>&1 | bunyan
+
+BUGS
+----
+
+DSA keys do not work when loaded via the SSH agent.
+
+Report bugs at [Github](https://github.com/joyent/node-manta/issues)
diff --git a/docs/man/mput.md b/docs/man/mput.md
index abf24b8..5088ba8 100644
--- a/docs/man/mput.md
+++ b/docs/man/mput.md
@@ -97,7 +97,7 @@ ENVIRONMENT
 -----------
 
 `MANTA_USER`
-  In place of `-a, --account`
+  In place of `-a, --account`.
 
 `MANTA_SUBUSER`
   In place of `--user`.
diff --git a/docs/man/mrm.md b/docs/man/mrm.md
index 6326f6c..0ee9411 100644
--- a/docs/man/mrm.md
+++ b/docs/man/mrm.md
@@ -63,7 +63,7 @@ ENVIRONMENT
 -----------
 
 `MANTA_USER`
-  In place of `-a, --account`
+  In place of `-a, --account`.
 
 `MANTA_SUBUSER`
   In place of `--user`.
diff --git a/docs/man/mrmdir.md b/docs/man/mrmdir.md
index 4e47ea8..d68e55d 100644
--- a/docs/man/mrmdir.md
+++ b/docs/man/mrmdir.md
@@ -55,7 +55,7 @@ ENVIRONMENT
 -----------
 
 `MANTA_USER`
-  In place of `-a, --account`
+  In place of `-a, --account`.
 
 `MANTA_SUBUSER`
   In place of `--user`.
diff --git a/docs/man/msign.md b/docs/man/msign.md
index 9a08f42..946ef0d 100644
--- a/docs/man/msign.md
+++ b/docs/man/msign.md
@@ -81,7 +81,7 @@ ENVIRONMENT
 -----------
 
 `MANTA_USER`
-  In place of `-a, --account`
+  In place of `-a, --account`.
 
 `MANTA_SUBUSER`
   In place of `--user`.
diff --git a/docs/man/muntar.md b/docs/man/muntar.md
index c12e5af..fcc7d8f 100644
--- a/docs/man/muntar.md
+++ b/docs/man/muntar.md
@@ -95,7 +95,7 @@ OPTIONS
   ENVIRONMENT
 -----------
 `MANTA_USER`
-  In place of `-a, --account`
+  In place of `-a, --account`.
 
 `MANTA_SUBUSER`
   In place of `--user`.
diff --git a/lib/client.js b/lib/client.js
index a029d1d..0896a4a 100644
--- a/lib/client.js
+++ b/lib/client.js
@@ -3303,7 +3303,7 @@ MantaClient.prototype.createUpload = function createUpload(p, opts, cb) {
     assert.string(p, 'path');
     assert.object(opts, 'options');
     assert.string(opts.account, 'opts.account');
-    assert.optionalObject(opts.headers, 'opts.headers');
+    assert.optionalObject(opts.targetObjectHeaders, 'opts.targetObjectHeaders');
     assert.optionalNumber(opts.copies, 'opts.copies');
     assert.optionalNumber(opts.size, 'opts.size');
     assert.optionalString(opts.md5, 'opts.md5');
@@ -3320,8 +3320,8 @@ MantaClient.prototype.createUpload = function createUpload(p, opts, cb) {
 
     // Headers to store with the object (not sent as headers on the request)
     var headers = {};
-    if (opts.headers) {
-        headers = jsprim.deepCopy(opts.headers);
+    if (opts.targetObjectHeaders) {
+        headers = jsprim.deepCopy(opts.targetObjectHeaders);
     }
 
     if (opts.copies) {
diff --git a/man/man1/mchattr.1 b/man/man1/mchattr.1
index 099b2c6..a3e4f78 100644
--- a/man/man1/mchattr.1
+++ b/man/man1/mchattr.1
@@ -55,7 +55,7 @@ Print debug output to stderr.  Repeat option to increase verbosity.
 .SH ENVIRONMENT
 .TP
 \fB\fCMANTA_USER\fR
-In place of \fB\fC\-a, \-\-account\fR
+In place of \fB\fC\-a, \-\-account\fR\&.
 .TP
 \fB\fCMANTA_SUBUSER\fR
 In place of \fB\fC\-\-user\fR\&.
@@ -89,5 +89,4 @@ $ mchattr \-v ~~/stor/foo 2>&1 | bunyan
 .PP
 DSA keys do not work when loaded via the SSH agent.
 .PP
-Report bugs at Github
-\[la]https://github.com/joyent/node-manta/issues\[ra]
+Report bugs at Github \[la]https://github.com/joyent/node-manta/issues\[ra]
diff --git a/man/man1/mchmod.1 b/man/man1/mchmod.1
index 29b384b..e90e298 100644
--- a/man/man1/mchmod.1
+++ b/man/man1/mchmod.1
@@ -55,7 +55,7 @@ Print debug output to stderr.  Repeat option to increase verbosity.
 .SH ENVIRONMENT
 .TP
 \fB\fCMANTA_USER\fR
-In place of \fB\fC\-a, \-\-account\fR
+In place of \fB\fC\-a, \-\-account\fR\&.
 .TP
 \fB\fCMANTA_SUBUSER\fR
 In place of \fB\fC\-\-user\fR\&.
diff --git a/man/man1/mfind.1 b/man/man1/mfind.1
index 8be9b73..d288812 100644
--- a/man/man1/mfind.1
+++ b/man/man1/mfind.1
@@ -93,7 +93,7 @@ Print debug output to stderr.  Repeat option to increase verbosity.
 .SH ENVIRONMENT
 .TP
 \fB\fCMANTA_USER\fR
-In place of \fB\fC\-a, \-\-account\fR
+In place of \fB\fC\-a, \-\-account\fR\&.
 .TP
 \fB\fCMANTA_SUBUSER\fR
 In place of \fB\fC\-\-user\fR\&.
diff --git a/man/man1/mget.1 b/man/man1/mget.1
index 994adcd..74c2eec 100644
--- a/man/man1/mget.1
+++ b/man/man1/mget.1
@@ -89,7 +89,7 @@ Print debug output to stderr.  Repeat option to increase verbosity.
 .SH ENVIRONMENT
 .TP
 \fB\fCMANTA_USER\fR
-In place of \fB\fC\-a, \-\-account\fR
+In place of \fB\fC\-a, \-\-account\fR\&.
 .TP
 \fB\fCMANTA_SUBUSER\fR
 In place of \fB\fC\-\-user\fR\&.
@@ -123,5 +123,4 @@ $ mget \-vv ~~/stor/foo 2>&1 | bunyan
 .PP
 DSA keys do not work when loaded via the SSH agent.
 .PP
-Report bugs at Github
-\[la]https://github.com/joyent/node-manta/issues\[ra]
+Report bugs at Github \[la]https://github.com/joyent/node-manta/issues\[ra]
diff --git a/man/man1/minfo.1 b/man/man1/minfo.1
index 0bbcd27..625af70 100644
--- a/man/man1/minfo.1
+++ b/man/man1/minfo.1
@@ -53,10 +53,10 @@ Print debug output to stderr.  Repeat option to increase verbosity.
 .SH ENVIRONMENT
 .TP
 \fB\fCMANTA_USER\fR
-In place of \fB\fC\-a, \-\-account\fR
+In place of \fB\fC\-a, \-\-account\fR\&.
 .TP
 \fB\fCMANTA_SUBUSER\fR
-In place of \fB\fC\-\-user\fR
+In place of \fB\fC\-\-user\fR\&.
 .TP
 \fB\fCMANTA_KEY_ID\fR
 In place of \fB\fC\-k, \-\-key\fR\&.
@@ -87,5 +87,4 @@ $ minfo \-vv ~~/stor/foo 2>&1 | bunyan
 .PP
 DSA keys do not work when loaded via the SSH agent.
 .PP
-Report bugs at Github
-\[la]https://github.com/joyent/node-manta/issues\[ra]
+Report bugs at Github \[la]https://github.com/joyent/node-manta/issues\[ra]
diff --git a/man/man1/mjob.1 b/man/man1/mjob.1
index e44666d..aed8749 100644
--- a/man/man1/mjob.1
+++ b/man/man1/mjob.1
@@ -345,10 +345,10 @@ Do not output any informative messages.
 .SH ENVIRONMENT
 .TP
 \fB\fCMANTA_USER\fR
-In place of \fB\fC\-a, \-\-account\fR
+In place of \fB\fC\-a, \-\-account\fR\&.
 .TP
 \fB\fCMANTA_SUBUSER\fR
-In place of \fB\fC\-\-user\fR
+In place of \fB\fC\-\-user\fR\&.
 .TP
 \fB\fCMANTA_KEY_ID\fR
 In place of \fB\fC\-k, \-\-key\fR\&.
diff --git a/man/man1/mln.1 b/man/man1/mln.1
index 0eff428..a6771f0 100644
--- a/man/man1/mln.1
+++ b/man/man1/mln.1
@@ -55,10 +55,10 @@ Print debug output to stderr.  Repeat option to increase verbosity.
 .SH ENVIRONMENT
 .TP
 \fB\fCMANTA_USER\fR
-In place of \fB\fC\-a, \-\-account\fR
+In place of \fB\fC\-a, \-\-account\fR\&.
 .TP
 \fB\fCMANTA_SUBUSER\fR
-In place of \fB\fC\-\-user\fR
+In place of \fB\fC\-\-user\fR\&.
 .TP
 \fB\fCMANTA_KEY_ID\fR
 In place of \fB\fC\-k, \-\-key\fR\&.
@@ -89,5 +89,4 @@ $ mln \-vv ~~/stor/foo 2>&1 | bunyan
 .PP
 DSA keys do not work when loaded via the SSH agent.
 .PP
-Report bugs at Github
-\[la]https://github.com/joyent/node-manta/issues\[ra]
+Report bugs at Github \[la]https://github.com/joyent/node-manta/issues\[ra]
diff --git a/man/man1/mlogin.1 b/man/man1/mlogin.1
index 85eb62f..74d5eb4 100644
--- a/man/man1/mlogin.1
+++ b/man/man1/mlogin.1
@@ -127,10 +127,10 @@ Don't print session establishment status messages.
 .SH ENVIRONMENT
 .TP
 \fB\fCMANTA_USER\fR
-In place of \fB\fC\-a, \-\-account\fR
+In place of \fB\fC\-a, \-\-account\fR\&.
 .TP
 \fB\fCMANTA_SUBUSER\fR
-In place of \fB\fC\-\-user\fR
+In place of \fB\fC\-\-user\fR\&.
 .TP
 \fB\fCMANTA_KEY_ID\fR
 In place of \fB\fC\-k, \-\-key\fR\&.
@@ -150,5 +150,4 @@ where \fB\fC:login\fR is the account login name.
 .PP
 DSA keys do not work when loaded via the SSH agent.
 .PP
-Report bugs at Github
-\[la]https://github.com/joyent/node-manta/issues\[ra]
+Report bugs at Github \[la]https://github.com/joyent/node-manta/issues\[ra]
diff --git a/man/man1/mls.1 b/man/man1/mls.1
index 431918e..d164339 100644
--- a/man/man1/mls.1
+++ b/man/man1/mls.1
@@ -69,7 +69,7 @@ Print debug output to stderr.  Repeat option to increase verbosity.
 .SH ENVIRONMENT
 .TP
 \fB\fCMANTA_USER\fR
-In place of \fB\fC\-a, \-\-account\fR
+In place of \fB\fC\-a, \-\-account\fR\&.
 .TP
 \fB\fCMANTA_SUBUSER\fR
 In place of \fB\fC\-\-user\fR\&.
@@ -103,5 +103,4 @@ $ mls \-vv ~~/stor 2>&1 | bunyan
 .PP
 DSA keys do not work when loaded via the SSH agent.
 .PP
-Report bugs at Github
-\[la]https://github.com/joyent/node-manta/issues\[ra]
+Report bugs at Github \[la]https://github.com/joyent/node-manta/issues\[ra]
diff --git a/man/man1/mmkdir.1 b/man/man1/mmkdir.1
index b57b029..2a284d2 100644
--- a/man/man1/mmkdir.1
+++ b/man/man1/mmkdir.1
@@ -54,7 +54,7 @@ Print debug output to stderr.  Repeat option to increase verbosity.
 .SH ENVIRONMENT
 .TP
 \fB\fCMANTA_USER\fR
-In place of \fB\fC\-a, \-\-account\fR
+In place of \fB\fC\-a, \-\-account\fR\&.
 .TP
 \fB\fCMANTA_SUBUSER\fR
 In place of \fB\fC\-\-user\fR\&.
@@ -88,5 +88,4 @@ $ mmkdir \-vv ~~/stor/foo 2>&1 | bunyan
 .PP
 DSA keys do not work when loaded via the SSH agent.
 .PP
-Report bugs at Github
-\[la]https://github.com/joyent/node-manta/issues\[ra]
+Report bugs at Github \[la]https://github.com/joyent/node-manta/issues\[ra]
diff --git a/man/man1/mmpu.1 b/man/man1/mmpu.1
new file mode 100644
index 0000000..af2a456
--- /dev/null
+++ b/man/man1/mmpu.1
@@ -0,0 +1,270 @@
+.TH mmpu 1 "June 2017" Manta "Manta Commands"
+.SH NAME
+.PP
+mmpu \- manage manta multipart uploads
+.SH SYNOPSIS
+.PP
+\fB\fCmmpu\fR [OPTION...] command [command\-specific arguments]
+.SH DESCRIPTION
+.PP
+mmpu allows you to interact with Manta's multipart upload API. The multipart
+upload API allows you to upload an object to Manta in chunks, or "parts",
+instead of through a single PUT request.  After all parts have been uploaded,
+you can "commit" your multipart upload, exposing your uploaded object as a
+single file in Manta.  Using mmpu, you can create multipart uploads, upload
+parts to them, and commit or "abort" (cancel) them. You may also list
+non\-garbage collected uploads and see which parts for a given upload.
+.PP
+The primary reference for a multipart upload is its UUID.  Most commands operate
+on multipart uploads by UUID.  Parts of a given upload are referenced by their
+part number.
+.PP
+The final object created by a committed multipart upload is referred to as the
+"target object" of the multipart upload.  Additionally, a multipart upload is
+often referred to as an "MPU" or just an "upload".
+.SH COMMON OPTIONS
+.PP
+The following options are supported in all commands:
+.TP
+\fB\fC\-a, \-\-account login\fR
+Authenticate as account (login name).
+.TP
+\fB\fC\-h, \-\-help\fR
+Print a help message and exit.
+.TP
+\fB\fC\-i, \-\-insecure\fR
+This option explicitly allows "insecure" SSL connections and transfers.  All
+SSL connections are attempted to be made secure by using the CA certificate
+bundle installed by default.
+.TP
+\fB\fC\-k, \-\-key fingerprint\fR
+Authenticate using the SSH key described by \fB\fCfingerprint\fR\&.  The key must
+either be in \fB\fC~/.ssh\fR or loaded in the SSH agent via \fB\fCssh\-add\fR\&.
+.TP
+\fB\fC\-\-role=ROLE,ROLE,...\fR
+Specify which roles to assume for the request.
+.TP
+\fB\fC\-\-user user\fR
+Authenticate as user under account.
+.TP
+\fB\fC\-u, \-\-url url\fR
+Manta base URL (such as \fB\fChttps://manta.us\-east.joyent.com\fR).
+.TP
+\fB\fC\-v, \-\-verbose\fR
+Print debug output to stderr.  Repeat option to increase verbosity.
+.SH COMMANDS
+.PP
+The following commands and options are supported:
+.SS create [OPTIONS...] PATH
+.PP
+Creates a multipart upload, that, when committed, will be exposed as a Manta
+object at PATH.  When this command completes successfully, it prints a
+multipart upload UUID to stdout.  You should save this UUID for use in other
+commands.
+.PP
+For example, if you wanted to create a multipart upload for which the target
+object will be stored at ~~/stor/foo.txt, you would run:
+.PP
+.RS
+.nf
+$ mmpu create ~~/stor/foo.txt
+.fi
+.RE
+.PP
+Since objects created through the multipart upload API should be
+indistinguishable from objects created through a normal Manta PUT, the \fB\fCcreate\fR
+command supports similar options that you might use for \fB\fCmput\fR\&.  For example,
+to create a multipart upload that has 3 copies of the target object, and stores
+some additional headers on the target object metadata, you would run:
+.PP
+.RS
+.nf
+$ mmpu create ~~/stor/foo.txt \-c 3 \-H 'content\-type: text/plain' \\
+                                   \-H 'access\-control\-allow\-origin: *'
+.fi
+.RE
+.PP
+You may also add additional data integrity checks that will be validated
+for your target object when you commit your upload.  For example, if you want
+to verify that your target object ~~/stor/foo.txt is 15728640 bytes and has
+the content\-md5 value "FLFyNOI3UFQhtkkrjXV1Bw==", you would run:
+.PP
+.RS
+.nf
+$ mmpu create ~~/stor/foo.txt \-s 15728650 \\
+                              \-m 'FLFyNOI3UFQhtkkrjXV1Bw=='
+.fi
+.RE
+.PP
+The following options are supported on \fB\fCcreate\fR:
+.TP
+\fB\fC\-c, \-\-copies num_copies\fR
+Specify the number of copies of the target object to store.
+.TP
+\fB\fC\-H, \-\-header header\fR
+Specify a HTTP header to store in the target object's metadata.  As with all
+Manta objects, you may store up to 4 KB of headers on the target object.
+.TP
+\fB\fC\-m, \-\-md5 md5\fR
+Validate the \fB\fCcontent\-md5\fR of the target object against the given md5 string.
+If the MD5 of the target object differs from the provided MD5 when the object
+is committed, the commit will fail.
+.IP
+Note that the behavior of this option is different the similar option for \fB\fCmput\fR\&.
+.TP
+\fB\fC\-s, \-\-size size\fR
+Validate the size of the target object against \fB\fCsize\fR\&.  If the size of the
+target object differs from this size when the object is committed, the commit
+will fail.
+.SS upload [\-f file] [OPTIONS...] MPU PART_NUM
+.PP
+Uploads the contents of stdin, or from the file specified with \fB\fC\-f\fR, to the
+specified part for a given upload.
+.PP
+Note that parts are zero\-indexed.  For example, if you wanted to upload the
+first part of upload be5f05d6\-7daa\-6869\-d19b\-c1f7fbcf6d8d, and the part's
+contents are stored locally at ./part0.txt, you could run the following
+equivalent commands:
+.PP
+.RS
+.nf
+$ cat part0.txt | mmpu upload be5f05d6\-7daa\-6869\-d19b\-c1f7fbcf6d8d 0
+.fi
+.RE
+.PP
+Or:
+.PP
+.RS
+.nf
+$ mmpu upload \-f part0.txt be5f05d6\-7daa\-6869\-d19b\-c1f7fbcf6d8d 0
+.fi
+.RE
+.PP
+When this command completes successfully, it prints the etag of the part
+that was uploaded to stdout.  You should save this etag to use when committing
+the upload.
+.PP
+The following options are supported on \fB\fCupload\fR:
+.TP
+\fB\fC\-f, \-\-file file\fR
+Upload the contents of \fB\fCfile\fR as the part, instead of the contents of stdin.
+.TP
+\fB\fC\-\-progress\fR
+Force the progress bar to draw, even when stderr is redirected.
+.TP
+\fB\fC\-q, \-\-quiet\fR
+Do not display a progress meter.
+.SS commit MPU [ETAG...]
+.PP
+Commits a given upload with the parts specified.  The parts are represented by
+their position in the argument list, which indicates their part number (with
+part 0 being the first argument after the upload ID, the second part the next
+argument after that, and so on), and their contents are represented by their
+etag.  If the etag provided for a part does not match the etag of the part when
+the commit process fetches its metadata, an error will be returned.
+.PP
+For example, if you want to commit an upload with id
+be5f05d6\-7daa\-6869\-d19b\-c1f7fbcf6d8d and the following parts:
+    part 0: etag f14a41dc\-7d28\-6fdb\-e07c\-d54c0adcdf35
+    part 1: etag e2893e52\-9ba3\-64fe\-eec9\-b4663835ad01
+    part 2: etag 73b44fa7\-fbd3\-efea\-b7d0\-cd8098e1d928
+.PP
+You would run:
+    $ mmpu commit be5f05d6\-7daa\-6869\-d19b\-c1f7fbcf6d8d \[rs]
+                  f14a41dc\-7d28\-6fdb\-e07c\-d54c0adcdf35 \[rs]
+                  e2893e52\-9ba3\-64fe\-eec9\-b4663835ad01 \[rs]
+                  73b44fa7\-fbd3\-efea\-b7d0\-cd8098e1d928
+.PP
+The multipart upload API does not require that you commit all parts uploaded,
+as long as zero or more consecutive parts are committed, and all parts meet
+part size restraints.  So the following commands committing the same upload as
+above are also valid, but would create a different target object than the
+first example.
+.PP
+This command would commit a zero\-byte target object:
+    $ mmpu commit be5f05d6\-7daa\-6869\-d19b\-c1f7fbcf6d8d
+.PP
+This would create an object consisting of only the first two parts:
+    $ mmpu commit be5f05d6\-7daa\-6869\-d19b\-c1f7fbcf6d8d \[rs]
+                  f14a41dc\-7d28\-6fdb\-e07c\-d54c0adcdf35 \[rs]
+                  e2893e52\-9ba3\-64fe\-eec9\-b4663835ad01
+.PP
+Once the commit process has begun for a given MPU, it cannot be aborted, or
+committed with a different set of parts than it was initially committed with,
+but you may retry the commit if needed.
+.PP
+To see the current status of an upload, use the \fB\fCget\fR command.
+.SS abort MPU
+.PP
+Aborts a multipart upload.
+.PP
+Once an upload has begun aborting, it may not be committed, but you may retry
+the abort operation if needed.  To see the current status of an upload, use the
+\fB\fCget\fR command.
+.SS get MPU
+.PP
+Fetches a JSON blob information about an upload, including its status: created,
+finalizing, or done (committed or aborted).  A finalizing upload is one
+that is either in the process of being committed or aborted, or a commit or
+abort request began but failed.
+.SS list
+.PP
+Lists all multipart uploads for a user (note, this can also be done with a
+normal \fB\fCmfind\fR call).  Note that these uploads may be in any state; their
+presence merely indicates they have not been garbage collected yet.
+.PP
+If you wish to list all parts that have been uploaded as well, you can run:
+.PP
+.RS
+.nf
+$ mmpu list \-p
+.fi
+.RE
+.PP
+To see the parts of only one upload, you should use the \fB\fCparts\fR command.
+.PP
+The following options are supported on \fB\fClist\fR:
+.TP
+\fB\fC\-p, \-\-includeParts\fR
+List parts in additional to uploads.
+.SS parts MPU
+.PP
+Lists all parts that have been uploaded to a given multipart upload.
+.SH ENVIRONMENT
+.TP
+\fB\fCMANTA_USER\fR
+In place of \fB\fC\-a, \-\-account\fR\&.
+.TP
+\fB\fCMANTA_SUBUSER\fR
+In place of \fB\fC\-\-user\fR\&.
+.TP
+\fB\fCMANTA_KEY_ID\fR
+In place of \fB\fC\-k, \-\-key\fR\&.
+.TP
+\fB\fCMANTA_ROLE\fR
+In place of \fB\fC\-\-role\fR\&.
+.TP
+\fB\fCMANTA_URL\fR
+In place of \fB\fC\-u, \-\-url\fR\&.
+.TP
+\fB\fCMANTA_TLS_INSECURE\fR
+In place of \fB\fC\-i, \-\-insecure\fR\&.
+.PP
+The shortcut \fB\fC~~\fR is equivalent to \fB\fC/:login\fR
+where \fB\fC:login\fR is the account login name.
+.SH DIAGNOSTICS
+.PP
+When using the \fB\fC\-v\fR option, diagnostics will be sent to stderr in bunyan
+output format.  As an example of tracing all information about a request,
+try:
+.PP
+.RS
+.nf
+$ mmpu create \-vv ~~/stor/foo 2>&1 | bunyan
+.fi
+.RE
+.SH BUGS
+.PP
+DSA keys do not work when loaded via the SSH agent.
+.PP
+Report bugs at Github \[la]https://github.com/joyent/node-manta/issues\[ra]
diff --git a/man/man1/mput.1 b/man/man1/mput.1
index d05265a..2c93436 100644
--- a/man/man1/mput.1
+++ b/man/man1/mput.1
@@ -96,7 +96,7 @@ Print debug output to stderr.  Repeat option to increase verbosity.
 .SH ENVIRONMENT
 .TP
 \fB\fCMANTA_USER\fR
-In place of \fB\fC\-a, \-\-account\fR
+In place of \fB\fC\-a, \-\-account\fR\&.
 .TP
 \fB\fCMANTA_SUBUSER\fR
 In place of \fB\fC\-\-user\fR\&.
diff --git a/man/man1/mrm.1 b/man/man1/mrm.1
index c2cdbb9..dc0a994 100644
--- a/man/man1/mrm.1
+++ b/man/man1/mrm.1
@@ -54,7 +54,7 @@ Print debug output to stderr.  Repeat option to increase verbosity.
 .SH ENVIRONMENT
 .TP
 \fB\fCMANTA_USER\fR
-In place of \fB\fC\-a, \-\-account\fR
+In place of \fB\fC\-a, \-\-account\fR\&.
 .TP
 \fB\fCMANTA_SUBUSER\fR
 In place of \fB\fC\-\-user\fR\&.
diff --git a/man/man1/mrmdir.1 b/man/man1/mrmdir.1
index 289947e..81651b2 100644
--- a/man/man1/mrmdir.1
+++ b/man/man1/mrmdir.1
@@ -46,7 +46,7 @@ Print debug output to stderr.  Repeat option to increase verbosity.
 .SH ENVIRONMENT
 .TP
 \fB\fCMANTA_USER\fR
-In place of \fB\fC\-a, \-\-account\fR
+In place of \fB\fC\-a, \-\-account\fR\&.
 .TP
 \fB\fCMANTA_SUBUSER\fR
 In place of \fB\fC\-\-user\fR\&.
@@ -80,5 +80,4 @@ $ mrmdir \-vv ~~/stor/foo 2>&1 | bunyan
 .PP
 DSA keys do not work when loaded via the SSH agent.
 .PP
-Report bugs at Github
-\[la]https://github.com/joyent/node-manta/issues\[ra]
+Report bugs at Github \[la]https://github.com/joyent/node-manta/issues\[ra]
diff --git a/man/man1/msign.1 b/man/man1/msign.1
index b075369..a8d9b72 100644
--- a/man/man1/msign.1
+++ b/man/man1/msign.1
@@ -75,7 +75,7 @@ Print debug output to stderr.  Repeat option to increase verbosity.
 .SH ENVIRONMENT
 .TP
 \fB\fCMANTA_USER\fR
-In place of \fB\fC\-a, \-\-account\fR
+In place of \fB\fC\-a, \-\-account\fR\&.
 .TP
 \fB\fCMANTA_SUBUSER\fR
 In place of \fB\fC\-\-user\fR\&.
@@ -109,5 +109,4 @@ $ msign \-vv ~~/stor/foo 2>&1 | bunyan
 .PP
 DSA keys do not work when loaded via the SSH agent.
 .PP
-Report bugs at Github
-\[la]https://github.com/joyent/node-manta/issues\[ra]
+Report bugs at Github \[la]https://github.com/joyent/node-manta/issues\[ra]
diff --git a/man/man1/muntar.1 b/man/man1/muntar.1
index fc49feb..80ec6fd 100644
--- a/man/man1/muntar.1
+++ b/man/man1/muntar.1
@@ -86,7 +86,7 @@ Print debug output to stderr.  Repeat option to increase verbosity.
 .SH   ENVIRONMENT
 .TP
 \fB\fCMANTA_USER\fR
-In place of \fB\fC\-a, \-\-account\fR
+In place of \fB\fC\-a, \-\-account\fR\&.
 .TP
 \fB\fCMANTA_SUBUSER\fR
 In place of \fB\fC\-\-user\fR\&.
diff --git a/package.json b/package.json
index d35566f..6e072ff 100644
--- a/package.json
+++ b/package.json
@@ -7,7 +7,7 @@
         "type": "git",
         "url": "git://github.com/joyent/node-manta.git"
     },
-    "version": "4.4.0",
+    "version": "4.4.1",
     "main": "./lib/index.js",
     "dependencies": {
         "assert-plus": "^1.0.0",
